### DISCUSSION TITLE
Use DataType.array in registerFunction

### DISCUSSION THREAD

**Discussion by imatveev** [oguimbal/pg-mem#99] [Q&A] [answered]
Trying to do something like:
```
	db.public.registerFunction({
		name: 'contains',
		args: [DataType.array, DataType.integer],
		returns: DataType.bool,
		implementation: (array, value) => array.contains(value)
	});
```
And getting:
`type "array" does not exist`
Is it possible somehow to use array as an argument?

**Comment by oguimbal** [comment, marked as answer]
Hi !

I wondered when someone would need this 😀

'array' is a type that is a bit special.

I exposed the necessary methods in `pg-mem@1.9.5` (which were there, but not exposed)... so you can do this like that:

```typescript
db.public.registerFunction({
		name: 'contains',
		args: [db.public.getType(DataType.integer).asArray(), DataType.integer],
		returns: DataType.bool,
		implementation: (array: number[], value: number) => array.contains(value)
	});
```

**Reply by imatveev** [reply]
@oguimbal That works like a charm! Thank you very much!
My use-case - testing a service that uses Presto as the data source.

P.S. After an update Schema started to require asArray implementation, maybe should be optional in type:
![image](https://user-images.githubusercontent.com/10616396/115452941-b9d37780-a227-11eb-9a6e-0e62985e1142.png)

**Reply by oguimbal** [reply]
Mmmh... I'm quite surprised that worked before ?

The `type` field is supposed to receive a type implementation, which implements `IType` (like the ones you get when calling `db.public.getType(...)`, not a type serialized definition (like `DataType.integer`).

I modified the `Schema` definition so you can use serialized as well:

```typescript
const schema: Schema = {
   name: table,
   fields: [ { type: DataType.integer, ...}]
}
```

👉 `pg-mem@1.9.6`
